src/rcc.rs

```diff
@@ -1,26 +1,341 @@
+    //! be configured and thus has a frequency component.
+    //!
+    //! To use them, compose them and feed them to, e.g., sysclk.
+    //!
+    //! ```rust
+    //! let mut rcc = RCC.constrain();
+    //! let msi_clk = clocking::MediumSpeedInternalRC::new(8_000_000, false);
+    //! let sys_clk_src = clocking::SysClkSource::MSI(msi_clk);
+    //! let cfgr = rcc.cfgr.sysclk(sys_clk_src);
+    //! ```
+    //!
+    //! The PLL is a bit more complex because it _is_ a source (`PLLClkOutput`) and also _requires_
+    //! a source (`PLLClkSource`), but you compose the types similarly.
+
+    use super::rcc;
+
+    pub trait InputClock {
+        fn freq(&self) -> u32;
+    }
+
+    /// High-speed internal 16 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedInternal16RC {
+        pub always_on: bool,
+        pub auto_start: bool,
+    }
+
+    impl HighSpeedInternal16RC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| w
+                          .hsion()
+                          .set_bit()
+                          .hsikeron()
+                          .bit(self.always_on)
+                          .hsiasfs()
+                          .bit(self.auto_start)
+                         );
+            while rcc.cr.read().hsirdy().bit_is_clear() {}
+            (16_000_000, 0b01)
+        }
+    }
+
+    /// Medium-speed internal 100 kHz - 48 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct MediumSpeedInternalRC {
+        f: u32,
+        auto_cal: bool,
+        bits: u8,
+    }
+
+    impl MediumSpeedInternalRC {
+
+        pub fn new(f: u32, a: bool) -> Self {
+            let bits = match f {
+                100_000 =>    0b0000,
+                200_000 =>    0b0001,
+                400_000 =>    0b0010,
+                800_000 =>    0b0011,
+                1_000_000 =>  0b0100,
+                2_000_000 =>  0b0101,
+                4_000_000 =>  0b0110,
+                8_000_000 =>  0b0111,
+                16_000_000 => 0b1000,
+                24_000_000 => 0b1001,
+                32_000_000 => 0b1010,
+                48_000_000 => 0b1011,
+                _ => panic!("bad MSI speed value!"),
+            };
+            MediumSpeedInternalRC {
+                f: f,
+                auto_cal: a,
+                bits: bits,
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
+            self.bits
+        }
+
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| unsafe {
+                w.msirange()
+                    .bits(self.bits())
+                    .msirgsel()
+                    .set_bit()
+            });
+            while rcc.cr.read().msirdy().bit_is_clear() {}
+
+            if self.auto_cal {
+                // FIXME This... may not work? I'm not sure if I've got a board problem or using
```

Comment:
Could it be it requires https://github.com/DoumanAsh/stm32l4x6_hal/blob/master/src/power.rs#L45 ?